velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp

```diff
@@ -0,0 +1,381 @@
+    return config_->get("hive.gcs.scheme", std::string("https"));
+  }
+
+  std::shared_ptr<gc::Credentials> credentials() const {
+    return credentials_;
+  }
+};
+
+class GCSFileSystem::Impl {
+ public:
+  Impl(const Config* config) : gcsConfig_(config) {}
+
+  ~Impl() = default;
+
+  // Use the input Config parameters and initialize the GCSClient.
+  void initializeClient() {
+    auto options = gc::Options{};
+    std::string scheme = gcsConfig_.scheme();
+    // if (scheme.empty()) scheme = "https";
+    if (scheme == "https") {
+      options.set<gc::UnifiedCredentialsOption>(
+          gc::MakeGoogleDefaultCredentials());
+    } else {
+      options.set<gc::UnifiedCredentialsOption>(gc::MakeInsecureCredentials());
+    }
+    options.set<gcs::UploadBufferSizeOption>(kUploadBufferSize);
+    if (!gcsConfig_.endpointOverride().empty()) {
+      options.set<gcs::RestEndpointOption>(
+          scheme + "://" + gcsConfig_.endpointOverride());
+    }
+    if (gcsConfig_.credentials()) {
+      options.set<gc::UnifiedCredentialsOption>(gcsConfig_.credentials());
+    }
+    client_ = std::make_shared<gcs::Client>(options);
+  }
+
+  std::shared_ptr<gcs::Client> getClient() const {
+    return client_;
+  }
+
+ private:
+  const GCSConfig gcsConfig_;
+  std::shared_ptr<gcs::Client> client_;
+};
+
+GCSFileSystem::GCSFileSystem(std::shared_ptr<const Config> config)
+    : FileSystem(config) {
+  impl_ = std::make_shared<Impl>(config.get());
+}
+
+void GCSFileSystem::initializeClient() {
+  impl_->initializeClient();
+}
+
+std::unique_ptr<ReadFile> GCSFileSystem::openFileForRead(
+    std::string_view path,
+    const FileOptions& /*unused*/) {
+  const std::string gcspath = gcsPath(path);
+  auto gcsfile = std::make_unique<GCSReadFile>(gcspath, impl_->getClient());
+  gcsfile->initialize();
+  return gcsfile;
+}
+
+std::unique_ptr<WriteFile> GCSFileSystem::openFileForWrite(
+    std::string_view path,
+    const FileOptions& /*unused*/) {
+  VELOX_NYI();
+}
+
+void GCSFileSystem::remove(std::string_view path) {
+  if (!isGCSFile(path)) {
+    VELOX_FAIL("File {} is not a valid gcs file", path);
+  }
+
+  // assumption it's a proper path
+  std::string bucket;
+  std::string object;
+  const std::string file = gcsPath(path);
+  setBucketAndKeyFromGCSPath(file, bucket, object);
+
+  if (!object.empty()) {
+    auto stat = impl_->getClient()->GetObjectMetadata(bucket, object);
+    if (!stat.ok()) {
+      checkGCSStatus(
+          stat.status(),
+          "Failed to get metadata for GCS object",
+          bucket,
+          object);
+    }
+  }
+  auto ret = impl_->getClient()->DeleteObject(bucket, object);
+  if (!ret.ok()) {
+    checkGCSStatus(
+        ret, "Failed to get metadata for GCS object", bucket, object);
+  }
+}
+
+bool GCSFileSystem::exists(std::string_view path) {
+  std::vector<std::string> result;
+  if (!isGCSFile(path))
+    VELOX_FAIL("File {} is not a valid gcs file", path);
+
+  // assumption it's a proper path
```

Comment:
// We assume 'path' is well-formed here.

Reply:
Done.